flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/PartitionableSourceITCase.scala

```diff
@@ -120,6 +127,33 @@ class PartitionableSourceITCase(
       )
     )
   }
+
+  @Test
+  def testPartitionPrunerCompileClassLoader(): Unit = {
+    val udfJavaCode =
+      s"""
+        |public class TrimUDF extends org.apache.flink.table.functions.ScalarFunction {
+        |   public String eval(String str) {
+        |     return str.trim();
+        |   }
+        |}
+        |""".stripMargin
+    val tmpDir: File = TEMPORARY_FOLDER.newFolder()
+    val udfJarFile: File = TestUserClassLoaderJar.createJarFile(
+      tmpDir, "flink-test-udf.jar", "TrimUDF", udfJavaCode)
+    val jars: util.List[URL] = new util.ArrayList[URL]()
+    jars.add(udfJarFile.toURI.toURL)
+    val cl = ClientUtils.buildUserCodeClassLoader(jars, java.util.Collections.emptyList(),
+      getClass.getClassLoader, new Configuration());
+    val ctx = TemporaryClassLoaderContext.of(cl)
+    tEnv.executeSql("create temporary function trimUDF as 'TrimUDF'");
+    checkResult("select * from MyTable where trimUDF(part1) = 'A' and part2 > 1",
+      Seq(
+        row(3, "Jack", "A", 2, 3)
+      )
+    )
+    ctx.close()
```

Comment:
Wrap this with `try... finally...` and close class loader context in `finally`. If the test fails the class loader will be changed and might cause other bugs in other tests.